test/parallel/test-inspector-connect-main-thread.js

```diff
@@ -0,0 +1,135 @@
+    (resolve) => worker.once('message', () => resolve(worker)));
+}
+
+function waitForWorkerMessages(worker) {
+  let workerDoneResolve, delayedConsoleLogResolve;
+  worker.on('message', ({ consoleLogIn10ms, messagesSent }) => {
+    if (consoleLogIn10ms) {
+      delayedConsoleLogResolve();
+    } else {
+      workerDoneResolve(messagesSent);
+    }
+  });
+  return [
+    new Promise((resolve) => { delayedConsoleLogResolve = resolve; }),
+    new Promise((resolve) => { workerDoneResolve = resolve; })
+  ];
+}
+
+function doConsoleLogIn5ms() {
+  setTimeout(() => {
+    console.log('Message for a test');
+  }, 5);
+}
+
+// This tests that inspector callbacks are called in a microtask
+// and do not interrupt the main code. Interrupting the code flow
+// can lead to unexpected behaviors.
+async function ensureListenerDoesNotInterrupt(session) {
+  const currentTime = Date.now();
+  let consoleLogHappened = false;
+  session.once('Runtime.consoleAPICalled',
+               () => { consoleLogHappened = true; });
+  parentPort.postMessage({ consoleLogIn10ms: true });
+  while ((Date.now() - currentTime) < 50) {
+    // Spin wait for 200ms
+  }
+  assert.strictEqual(consoleLogHappened, false);
+  await new Promise(queueMicrotask);
+  assert.strictEqual(consoleLogHappened, true);
+}
+
+async function RunTestWorker() {
+  const worker = await WaitForWorker();
+  const [consoleLogExpected, workerDonePromise] =
+    waitForWorkerMessages(worker);
+  consoleLogExpected.then(doConsoleLogIn5ms);
+  assert.strictEqual(toDebug(), 400);
+  assert.deepStrictEqual(await workerDonePromise, [
+    'Debugger.enable',
+    'Runtime.enable',
+    'Debugger.setBreakpointByUrl',
+    'Debugger.evaluateOnCallFrame',
+    'Debugger.resume'
+  ]);
+}
+
+async function RunDebugging() {
+  // Ensures the worker does not terminate too soon
+  parentPort.on('message', () => { });
+  const session = new Session();
+  session.connectToMainThread();
+  await post(session, 'Debugger.enable');
+  await post(session, 'Runtime.enable');
+  await post(session, 'Debugger.setBreakpointByUrl', {
+    'lineNumber': 11,
+    'url': pathToFileURL(path.resolve(__dirname, __filename)).toString(),
+    'columnNumber': 0,
+    'condition': ''
+  });
+  const pausedPromise = waitForNotification(session, 'Debugger.paused');
+  parentPort.postMessage('Ready');
+  const callFrameId = (await pausedPromise).params.callFrames[0].callFrameId;
+
+  // Delay to ensure main thread is truly suspended
+  await new Promise((resolve) => setTimeout(resolve, 50));
+
+  const { result: { value } } =
+      await post(session,
+                 'Debugger.evaluateOnCallFrame',
+                 { callFrameId, expression: 'a * 100' });
+  assert.strictEqual(value, 100);
+  await post(session, 'Debugger.resume');
+  await ensureListenerDoesNotInterrupt(session);
+  parentPort.postMessage({ messagesSent });
+  parentPort.close();
+  console.log('Worker is done');
+}
+
+if (isMainThread) {
+  RunTestWorker();
```

Comment:
```suggestion
  runTestWorker().then(common.mustCall());
```

(Adding “Changes requested” because the test really does not work in its current form)

Reply:
Done.